checker/tests/resourceleak-collections/CollectionOwnershipBasicTyping.java

```diff
@@ -0,0 +1,136 @@
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import org.checkerframework.checker.calledmethods.qual.*;
+import org.checkerframework.checker.collectionownership.qual.*;
+import org.checkerframework.checker.mustcall.qual.*;
+
+/*
+ * Test whether the defaults of resource collection fields, parameters, return types and new
+allocations
+ * are as expected.
+ */
+class CollectionOwnershipBasicTyping {
+
+  int n = 10;
+
+  /*
+   * Check that this return is allowed. The local list defaults to @OwningCollection, which
+   * the return type does too. Thus, return value is consistent with the return type.
+   */
+  Collection<Socket> checkReturn() {
+    List<Socket> list = new ArrayList<>();
+    return list;
+  }
+
+  /*
+   * Check that this return is disallowed. The parameter list defaults to @NotOwningCollection,
+   * which is a supertype of the @OwningCollection return type.
+   * Thus, return value is inconsistent with the return type.
+   */
+  Collection<Socket> checkReturn(List<Socket> list) {
+    // :: error: return
+    return list;
+  }
+
+  // this return looks harmless. However, it returns a collection, but not its ownership.
+  // Since no obligation will be created at call-site (it expects a non-owning collection),
+  // this method must ensure the returned collection has no obligations, which it fails to
+  // do in this case.
+  @NotOwningCollection
+  Collection<Socket> checkIllegalNotOwningReturn() {
+    List<Socket> l = new ArrayList<>();
+    // :: error: unfulfilled.collection.obligations
+    l.add(new Socket());
+    return l;
+  }
+
+  // this is the correct version of above. It passes ownership to another method first,
+  // and then returns the non-owning reference.
+  @NotOwningCollection
+  Collection<Socket> checkLegalNotOwningReturn() {
+    List<Socket> list = new ArrayList<>();
+    closeElements(list);
```

Comment:
This test case seems overly restrictive, based on my understanding of the type system's rules. In fact, you could have two tests here:
```
  @NotOwningCollection
  Collection<Socket> checkLegalNotOwningReturn1() {
    List<Socket> l = new ArrayList<>();
    l.add(new Socket());
    closeElements(l);
    return l;
  }

  @NotOwningCollection
  Collection<Socket> checkLegalNotOwningReturn2() {
    List<Socket> l = new ArrayList<>();
    return l; // OK, because `l` is empty and so should be @OwningCollectionNoObligations
  }
```